### ISSUE TITLE
Ontology Version IRI not absolute: Repair

### ISSUE THREAD

**Issue by matentzn** [owlcs/argo#3]
Almost all ontologies violate the fact that Version IRIs need to be absolute. Fixing it however causes another problem with the default repair strategy in the OWL API (which replaces the entire Ontology ID): If more than one ontologies in the imports closure have that problem, all will get the same ID. Which is again a (even more serious) violation.

@ignazio1977

**Comment by matentzn** [comment]
I think I would vote for a random UUID or something for the fix, or else some other that ensures that each ontology in the closure fixed that way is unique.

**Comment by matentzn** [comment]
Suggestion:
list(new SetOntologyID(ontology, new OWLOntologyID(Optional.of(IRI.create("urn:ontology#"+UUID.randomUUID())),
		            Optional.of(IRI.create("urn:ontology#"+UUID.randomUUID())))));

**Comment by matentzn** [comment]
@ignazio1977 This illustrates the problem: [BioportalRpubs](http://rpubs.com/matentzn/bioportal2017_03_30). You can see that most ontologies fail only on small accounts, and this non-absolute-version violation is by far the most extreme.

**Comment by alanruttenberg** [comment]
This smells of magic repairs. See https://github.com/owlcs/argo/issues/2#issuecomment-291205736 

Assuming that this error is caught, I think there should be more than one offer for repair. For example, the OBO ontologies have a policy on ontologyIRI and versionIRI and if a repair was to be made, then it should follow that [policy](http://obofoundry.org/id-policy.html).  Certainly a UUID as fix would be inappropriate for an OBO ontology.

**Comment by alanruttenberg** [comment]
BTW, the link  OntologyVersionIRINotAbsolute on [BioportalRpubs](http://rpubs.com/matentzn/bioportal2017_03_30) gives me a blank page.

FIrst suggestion for that page: If giving notice of a non-absolute Ontology/VersionIRI, display also what the current value is.

**Comment by matentzn** [comment]
You are right in that it is a magic repair. We probably need to distinguish the purpose of the repair. I am naturally interested only in violations relevant to reasoning, but if your goal is profile adherence ontology maintenance, you are of course right. I would inject UUIDs only to make the violation go away, in order to produce a serialised ontology that adheres the OWL 2 DL so that all reasoners have a fair chance to deal with it the same way. Tools should, as you say, point out the violation, and suggest to fix it by actually adding the Ontology Version IRI in the way the developers see fit.

**Comment by matentzn** [comment]
@ignazio1977 Given the extreme prevalence of this error, is it possible this is actually CAUSED by the OWL XML serialisation? It is hard to image that 300+ out of 400 ontologies have version IRIs.. I guess we all agree that non-absolute IRIs [are a violation](https://www.w3.org/TR/owl2-syntax/#Ontology_IRI_and_Version_IRI) are a violation, and should be reported. Is there any hope of designing a mechanism that gives a a default repair as a kind of "quick fix", and a slightly more elaborate one that is configurable by the user?

**Comment by matentzn** [comment]
@alanruttenberg The link actually works... Its rpubs that is broken, and in order to click on it you need to open it in a new tab.. :P I will for the next iteration give examples of that error..

**Comment by ignazio1977** [comment]
> is it possible this is actually CAUSED by the OWL XML serialisation?

I guess it's possible, haven't tried saving an ontology with an absolute version IRI to see if it comes out wrong.

**Comment by matentzn** [comment]
Sit tight. I will do some more analyses and come back to you.

**Comment by matentzn** [comment]
We should put this to rest. This was an issue that resulted from creating a new ontology from an old one, and does not seem an otherwise occurring problem. I will open a new issue.

**Comment by matentzn** [comment]
maybe shouldnt be closed yet as issue remains..
